### ISSUE TITLE
With strict type checking, improper overload detection with Iterable

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#3780]
**Describe the bug**
If you use strict type checking, and have a class that implements `__iter__()`, then having an overload match that class and also `Iterable` reports an error.

**To Reproduce**
```python
from typing import Iterable, overload, Iterator, Union


class Foo:
    def __iter__(self) -> Iterator[float]:
        ...


@overload
def apply2(func: Foo) -> str:
    ...


@overload
def apply2(func: Iterable[float]) -> int:
    ...


def apply2(func: Iterable[float]) -> Union[str, int]:
    return 5
```

reports error:
```text
callableover.py:10:5 - error: Overload 1 for "apply2" overlaps overload 2 and returns an incompatible type (reportOverlappingOverload)
```

**Expected behavior**
No error.

**Screenshots or Code**
See above.

**VS Code extension or command-line**
pyright command line with strict type checking set in `pyproject.toml` .  Version 1.1.265

**Additional context**

Came up in a pandas PR.

**Comment by erictraut** [comment]
This error is correct. As it says, overload 1 overlaps overload 2 and returns an incompatible type. `Foo` is compatible with `Iterable[float]`, so consider the following code.

```python
def func(x: Iterable[float]):
    reveal_type(apply2(x)) # int

func(Foo())
```

Mypy generates the same error as pyright in this case.

**Comment by erictraut** [comment]
I would recommend against designing a function like this in the first place. But if the function already exists and you're trying to find the best way to annotate it, you could do one of the following:
1. Stick with the solution you have. Add a `# type: ignore` to suppress the error, and live with the potential types errors (both false positive and negative) that result from code like my snippet above.
2. Change the return type of the second overload to `str | int` to reflect the fact that the input parameter might be a `Foo`, since `Foo` is also compatible with `Iterable[float]`. This closes the typing hole, but it means that users of the stub may need to add casts or additional runtime checks to validate that the return type is indeed an `int` rather than a `str`.

**Comment by erictraut** [comment]
In your example above, you haven't defined `Sub1`, `Sub2` or `Base`. I presume that `Sub1` and `Sub2` are peer classes that both derive from `Base`?

```python
class Base: ...
class Sub1(Base): ...
class Sub2(Base): ...
```

In your example, the overloads are applied in the manner you are hoping. However, here's a slight variant on your example where the overloads produce results contrary to what you are hoping.

```python
def func1(b: Base):
    reveal_type(apply3(b))

func1(Sub1())
```

**Comment by Dr-Irv** [comment]
Oops - copy/paste error:
```python
from typing import Union, overload
class Base:
    ...


class Sub1(Base):
    ...


class Sub2(Base):
    ...


@overload
def apply3(func: Sub1) -> str:
    ...


@overload
def apply3(func: Base) -> int:
    ...


def apply3(func: Base) -> Union[str, int]:
    ...


reveal_type(apply3(Sub1()))
reveal_type(apply3(Sub2()))
reveal_type(apply3(Base()))


def func1(b: Base):
    reveal_type(apply3(b))
```
produces what I expect:
```text
  callableover.py:41:5 - error: Overload 1 for "apply3" overlaps overload 2 and returns an incompatible type (reportOverlappingOverload)
  callableover.py:54:13 - information: Type of "apply3(Sub1())" is "str"
  callableover.py:55:13 - information: Type of "apply3(Sub2())" is "int"
  callableover.py:56:13 - information: Type of "apply3(Base())" is "int"
  callableover.py:60:17 - information: Type of "apply3(b)" is "int"
```

The matching is occurring as expected.  When `Sub1()` is the argument, result is `str`.  Any other subclass or the base class `Base`, produces `int`.

And from a typing perspective, calling `func1(Sub1())` should return `int`, because `func1` declared that it is accepting any `Base`.

So I'd still like to be able to ignore this "overlapping" message, because when objects of that type are passed statically, you get the result you expect.

**Comment by Dr-Irv** [comment]
@erictraut - not sure you saw my last response.  Appreciate the conversation.

I guess the question is this.  If you have a set of 3 overloads for a function that has an argument, and the type of the argument is wider in the second overload than in the first, and wider in the third overload than the second, and all 3 overloads return different types as a return type, then we know that the type checker will match them in order and return the appropriate result given the hierarchy of types of the arguments.  So are they really "overlapping", considering they will be matched in order?

**Comment by erictraut** [comment]
> So I'd still like to be able to ignore this "overlapping" message

Yes, this is a type error. These are really overlapping, and it can lead to problems, so pyright is correct to report an error here. As with any error, you are welcome to suppress it or ignore it.
